adage/analyze/models.py

```diff
@@ -0,0 +1,252 @@
+import re
+from django.db import models
+from django.core.exceptions import ValidationError, ObjectDoesNotExist
+from organisms.models import Organism
+from genes.models import Gene
+
+
+def validate_pyname(value):
+    """
+    Raise a ValidationError if value is not a valid Python name. See Python
+    docs at:
+    https://docs.python.org/2/reference/lexical_analysis.html#identifiers
+    for full specification.
+    """
+    if not re.match(r'[A-Za-z_][A-Za-z0-9_]*', value):
+        raise ValidationError(
+            "%(value)s is not a valid Python identifier",
+            params={'value': value},
+        )
+
+
+class Experiment(models.Model):
+    accession = models.CharField(max_length=48, primary_key=True)
+    name = models.CharField(max_length=1000)
+    description = models.TextField()
+
+    def __str__(self):
+        return self.accession
+
+# TODO: implement absolute urls for Experiments. see
+#  https://docs.djangoproject.com/en/1.8/ref/models/instances/#get-absolute-url
+# TODO: implement a model for samples uploaded by users
+
+
+class Sample(models.Model):
+    name = models.CharField(
+        "sample name",
+        max_length=80,
+        blank=False)
+    ml_data_source = models.CharField(
+        "Machine Learning data used for modeling, e.g. CEL file",
+        max_length=120,
+        null=True,
+        unique=True,
+        blank=True)
+    experiments = models.ManyToManyField(Experiment)
+
+    def __str__(self):
+        return "%d (%s)" % (self.id, self.name)
+
+    def get_annotation_dict(self):
+        return SampleAnnotation.objects.get_as_dict(self)
+
+    def get_annotation_items(self):
+        return iter(SampleAnnotation.objects.get_as_dict(self).items())
+
+
+class AnnotationTypeManager(models.Manager):
+    def create(self, typename, description=''):
+        ann_type = AnnotationType(
+            typename=typename,
+            description=description,
+        )
+        ann_type.full_clean()
+        ann_type.save()
+        return ann_type
+
+    def get_or_create(self, typename, description=''):
+        try:
+            ann_type = self.get(typename=typename)
+            created = False
+        except ObjectDoesNotExist:
+            ann_type = self.create(
+                typename=typename,
+                description=description)
+            created = True
+        return (ann_type, created)
+
+
+class AnnotationType(models.Model):
+    typename = models.CharField(
+        "name for this AnnotationType (usable as a Python identifier)",
+        max_length=40,
+        unique=True,
+        blank=False,
+        validators=[validate_pyname])
```

Comment:
I believe I added this validator before I understood that it isn't automatically run unless the model is used in a Django form, so this (and the `validate_pyname` function defined above) may be dead code that never gets executed.

Perhaps this has changed in Django 2.2 or maybe the REST framework will magically invoke it somehow. We should check that to confirm before removing it.

Reply:
Ok. For now, I'm just glad that at least you also know this might be an unevaluated code path.